### ISSUE TITLE
Using nested transactions with save points

### ISSUE THREAD

**Issue by thetutlage** [tgriesser/knex#3389]
# Environment

Knex version: `0.19.1`
Database + version: `PostgreSQL 11`
OS: `MacOS Mojave`

# Bug
I am trying to use nested transactions with save points and it doesn't seems to be working with promises API.

Here's some code (I am sharing the minimum required code to reproduce the problem)

**Following doesn't work**

```js
async function run () {
  const trx = await knex.transaction()
  const trx2 = await trx.transaction()
  console.log('never reached here')
}
```

**Following works**

```js
function runWithCallbacks () {
  knex.transaction((trx) => {
   trx.transaction((trx2) => {
    console.log('reached here>')
   })
 })
}
```

The first example doesn't work since the promise doesn't resolve unless `trx2` commit/rollback and it cannot, since the code never moves to the next line.

Is this the intended behavior?

**Comment by thetutlage** [comment]
I have been trying to look into the source code and there is hardly anything I am able to find that helps finding the root cause. All I have is the debug output.

```sh
  knex:tx trx1: Starting top level transaction +0ms
  knex:client acquired connection from pool: __knexUid2 +0ms
  knex:query BEGIN; trx1 +0ms
  knex:bindings undefined trx1 +0ms
  knex:tx trx3: Starting nested transaction +25ms
  knex:query SAVEPOINT trx3; trx3 +3ms
  knex:bindings undefined trx3 +3ms
```

The debug output is same with `await` calls and with callbacks.

**Comment by thetutlage** [comment]
Any chance someone can look into it?

**Comment by thetutlage** [comment]
@tgriesser, @elhigu, @kibertoad. Can someone please share some insights related to this issue? Right now I am not even 100% sure, if this is something expected to work or not.

**Comment by kibertoad** [comment]
@thetutlage Theoretically it should work. I'll take a look; sorry for the delay.

**Comment by kibertoad** [comment]
@thetutlage Here is what was happening. `trx.transaction` is a synchronous operation that expects you to pass transaction handler as a parameter; it does not support the Promise API the way Knex does. Hence correct way to write the above would be
```
  const trx = await knex.transaction()
   trx.transaction((trx2) => {
    console.log('reached here>')
   })
```

That said, it is confusing and inconsistent, hence see the PR raised to improve this behaviour (review and comments would be most welcome).

**Comment by thetutlage** [comment]
@kibertoad Thanks a lot for checking it out :). The PR does looks fine to me, but I may not the best person to judge that.

From the consistent API point of view, I also expected it work.

**Comment by thetutlage** [comment]
Found one regression [here](https://github.com/tgriesser/knex/blob/6ddb9fe0a2316f1018277e82ed6330272cc94c20/lib/transaction.js#L22). Now, since we are faking the container, the rejection behavior of rollback is to always reject.

**Comment by thetutlage** [comment]
Infact, transactions are quite broken between different dialects. The query handler from the base transaction class handles the use cases little bit differently.

## [Base Transaction](https://github.com/tgriesser/knex/blob/6ddb9fe0a2316f1018277e82ed6330272cc94c20/lib/transaction.js#L186)

Base transaction class checks the `sql` against a regex, which also handles the `savepoint` use case.

However, the following dialects are checking against a literal value or there is no `sql` check at all.

- [mssql](https://github.com/tgriesser/knex/blob/6ddb9fe0a2316f1018277e82ed6330272cc94c20/lib/dialects/mssql/transaction.js#L36)

- [mysql](https://github.com/tgriesser/knex/blob/6ddb9fe0a2316f1018277e82ed6330272cc94c20/lib/dialects/mysql/transaction.js#L35)

- [mysql2](https://github.com/tgriesser/knex/blob/6ddb9fe0a2316f1018277e82ed6330272cc94c20/lib/dialects/mysql2/transaction.js#L34)

- [oracle](https://github.com/tgriesser/knex/blob/6ddb9fe0a2316f1018277e82ed6330272cc94c20/lib/dialects/oracle/transaction.js#L32)

- [oracledb](
https://github.com/tgriesser/knex/blob/6ddb9fe0a2316f1018277e82ed6330272cc94c20/lib/dialects/oracledb/transaction.js#L37)

- `postgres` and `redshift` uses base transaction.

**Comment by kibertoad** [comment]
@thetutlage Would you consider contributing a fix?

**Comment by thetutlage** [comment]
Yeah sure, why not. You want me to also take the changes from your recent PR, or should we keep this separate

**Comment by kibertoad** [comment]
I'll update the PR with your proposed enhancement, so remaining part would be savepoint handling, which could be separate PR (and I assume we need more tests for it, if that remained unnoticed).

**Comment by thetutlage** [comment]
Sure, I will do a PR to make all dialects behave the same way for transactions and will see, if we can abstract the code to a helper function for consistency

**Comment by thetutlage** [comment]
Hello @kibertoad here's the PR. https://github.com/tgriesser/knex/pull/3399

**Comment by kibertoad** [comment]
@thetutlage Released in 0.19.2
